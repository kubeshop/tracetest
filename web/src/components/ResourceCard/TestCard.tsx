import {DownOutlined, RightOutlined} from '@ant-design/icons';
import {useMemo} from 'react';
import CreateButton from 'components/CreateButton';
import TestRunCard from 'components/RunCard/TestRunCard';
import TracetestAPI from 'redux/apis/Tracetest';
import {ResourceType} from 'types/Resource.type';
import Test from 'models/Test.model';
import TestRun from 'models/TestRun.model';
import Box from './Box';
import * as S from './ResourceCard.styled';
import ResourceCardActions from './ResourceCardActions';
import ResourceCardRuns from './ResourceCardRuns';
import ResourceCardSummary from './ResourceCardSummary';
import useRuns from './useRuns';

const {useLazyGetRunListQuery} = TracetestAPI.instance;

interface IProps {
  onEdit(id: string, lastRunId: number, type: ResourceType): void;
  onDelete(id: string, name: string, type: ResourceType): void;
  onRun(test: Test, type: ResourceType): void;
  onDuplicate(test: Test, type: ResourceType): void;
  onViewAll(id: string, type: ResourceType): void;
  test: Test;
}

const TestCard = ({onEdit, onDelete, onDuplicate, onRun, onViewAll, test}: IProps) => {
  const queryParams = useMemo(() => ({take: 5, testId: test.id}), [test.id]);
  const {isCollapsed, isLoading, list, onClick} = useRuns<TestRun, {testId: string}>(
    useLazyGetRunListQuery,
    queryParams
  );

  const shouldEdit = test.summary.hasRuns;
  const lastRunId = test.summary.runs; // assume the total of runs as the last run

  return (
    <S.Container $type={ResourceType.Test}>
      <S.TestContainer onClick={onClick}>
        {isCollapsed ? <RightOutlined data-cy={`collapse-test-${test.id}`} /> : <DownOutlined />}

        <Box num={test.summary.runs} type={ResourceType.Test} />

        <S.TitleContainer>
          <S.Title level={3}>{test.name}</S.Title>
          <S.Text>
            {test.trigger.method} {test.trigger.entryPoint && `â€¢ ${test.trigger.entryPoint}`}
          </S.Text>
        </S.TitleContainer>

        <ResourceCardSummary summary={test.summary} />

        <S.Row $gap={12}>
          {Test.shouldAllowRun(test.trigger.type) && (
            <CreateButton
              data-cy={`test-run-button-${test.id}`}
              ghost
              onClick={event => {
                event.stopPropagation();
                onRun(test, ResourceType.Test);
              }}
              type="primary"
            >
              Run
            </CreateButton>
          )}
          <ResourceCardActions
            id={test.id}
            shouldEdit={shouldEdit}
            onDelete={() => onDelete(test.id, test.name, ResourceType.Test)}
            onEdit={() => onEdit(test.id, lastRunId, ResourceType.Test)}
            onDuplicate={() => onDuplicate(test, ResourceType.Test)}
          />
        </S.Row>
      </S.TestContainer>

      <ResourceCardRuns
        hasMoreRuns={list.length === 5}
        hasRuns={Boolean(list.length)}
        isCollapsed={isCollapsed}
        isLoading={isLoading}
        onViewAll={() => onViewAll(test.id, ResourceType.Test)}
        resourcePath={`/test/${test.id}`}
      >
        <S.RunsListContainer data-cy="run-card-list">
          {list.map(run => (
            <TestRunCard key={run.id} linkTo={`/test/${test.id}/run/${run.id}`} run={run} testId={test.id} />
          ))}
        </S.RunsListContainer>
      </ResourceCardRuns>
    </S.Container>
  );
};

export default TestCard;
