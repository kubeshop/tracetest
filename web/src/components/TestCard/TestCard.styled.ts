import {MoreOutlined} from '@ant-design/icons';
import {Button, Typography} from 'antd';
import styled from 'styled-components';

import emptyStateIcon from 'assets/SpanAssertionsEmptyState.svg';

export const TestCard = styled.div<{$isCollapsed: boolean}>`
  box-shadow: 0 4px 8px rgba(153, 155, 168, 0.1);
  background: ${({theme}) => theme.color.white};
  border-left: ${({$isCollapsed, theme}) => $isCollapsed && `2px solid ${theme.color.primary}`};
  border-radius: 2px;
`;

export const InfoContainer = styled.div`
  cursor: pointer;
  display: grid;
  align-items: center;
  grid-template-columns: 20px 1fr 60px 2fr 220px 100px 20px;
  gap: 24px;
  padding: 16px 24px;
`;

export const ResultListContainer = styled.div`
  margin: 0 68px 54px 54px;
`;

export const TextContainer = styled.div`
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
`;

export const ButtonContainer = styled.div`
  display: flex;
  justify-content: flex-end;
`;

export const NameText = styled(Typography.Title).attrs({ellipsis: true, level: 3})`
  && {
    margin: 0;
  }
`;

export const Text = styled(Typography.Text)``;

export const ActionButton = styled(MoreOutlined)`
  color: ${({theme}) => theme.color.textSecondary};
  cursor: pointer;
  font-size: ${({theme}) => theme.size.lg};
`;

export const TestDetails = styled.div`
  text-align: right;
  width: 100%;
  margin-top: 8px;
`;

export const TestDetailsLink = styled(Button).attrs({
  type: 'link',
})`
  color: ${({theme}) => theme.color.primary};
  font-weight: 600;
  padding: 0;
`;

export const EmptyStateIcon = styled.img.attrs({
  src: emptyStateIcon,
})``;

export const EmptyStateContainer = styled.div`
  align-items: center;
  display: flex;
  flex-direction: column;
  gap: 14px;
  justify-content: center;
  margin: 10px 0 20px;
`;
