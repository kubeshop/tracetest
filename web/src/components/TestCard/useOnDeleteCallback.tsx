import {MenuInfo} from 'rc-menu/lib/interface';
import {useCallback} from 'react';

type Return = (info: Partial<MenuInfo>) => void;

export function useOnDeleteCallback(onDelete: () => void): Return {
  return useCallback(
    ({domEvent}) => {
      domEvent?.stopPropagation();
      onDelete();
    },
    [onDelete]
  );
}
