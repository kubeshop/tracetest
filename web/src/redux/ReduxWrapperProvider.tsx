// test-utils.jsx
import * as React from 'react';
import {Provider} from 'react-redux';
import {store} from './store';

export const ReduxWrapperProvider: React.FC = ({children}) => {
  return <Provider store={store}>{children}</Provider>;
};
