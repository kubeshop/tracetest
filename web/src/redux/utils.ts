import {PromiseWithKnownReason} from '@reduxjs/toolkit/dist/query/core/buildMiddleware/types';
import webSocketGateway, {IListenerFunction} from 'gateways/WebSocket.gateway';

export type {IListenerFunction} from 'gateways/WebSocket.gateway';

interface IInitWebSocketSubscription {
  listener: IListenerFunction;
  resource: string;
  waitToCleanSubscription: Promise<void>;
  waitToInitSubscription: PromiseWithKnownReason<any, any>;
}

/**
 * Implements RTK Query logic to receive streaming updates for persistent queries.
 * It allows a query to establish a WebSocket subscription, and apply updates to
 * the cached data when the information is received from the server.
 * For more information please take a look: https://redux-toolkit.js.org/rtk-query/usage/streaming-updates
 */
export async function initWebSocketSubscription({
  listener,
  resource,
  waitToCleanSubscription,
  waitToInitSubscription,
}: IInitWebSocketSubscription) {
  try {
    await waitToInitSubscription;
    webSocketGateway.subscribe(resource, listener);
  } catch {
    // no-op in case `waitToCleanSubscription` resolves before `waitToInitSubscription`,
    // in which case `waitToInitSubscription` will throw
  }
  await waitToCleanSubscription;
  webSocketGateway.unsubscribe(resource, '1234');
}
