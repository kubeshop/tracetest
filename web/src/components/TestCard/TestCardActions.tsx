import {Dropdown, Menu} from 'antd';
import * as S from './TestCard.styled';
import {useOnDeleteCallback} from './useOnDeleteCallback';

interface IProps {
  testId: string;

  onDelete(): void;
}

const TestCardActions: React.FC<IProps> = ({testId, onDelete}) => {
  const onClick = useOnDeleteCallback(onDelete);
  return (
    <Dropdown
      overlay={<Menu items={[{key: 'delete', label: <p data-cy="delete">Delete</p>, onClick}]} />}
      placement="bottomLeft"
      trigger={['click']}
    >
      <span data-cy={`test-actions-button-${testId}`} className="ant-dropdown-link" onClick={e => e.stopPropagation()}>
        <S.ActionButton />
      </span>
    </Dropdown>
  );
};

export default TestCardActions;
