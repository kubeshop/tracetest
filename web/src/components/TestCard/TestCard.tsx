import {DownOutlined, RightOutlined} from '@ant-design/icons';
import {Skeleton, Tooltip} from 'antd';
import {useCallback, useState} from 'react';

import {useLazyGetRunListQuery} from 'redux/apis/TraceTest.api';
import TestAnalyticsService from 'services/Analytics/TestAnalytics.service';
import {TTest} from 'types/Test.types';
import Date from 'utils/Date';
import * as S from './TestCard.styled';
import TestCardActions from './TestCardActions';
import ResultCardList from '../RunCardList/RunCardList';

interface IProps {
  onDelete(test: TTest): void;
  onRun(id: string): void;
  onViewAll(id: string): void;
  test: TTest;
}

const TestCard = ({onDelete, onRun, onViewAll, test}: IProps) => {
  const [isCollapsed, setIsCollapsed] = useState(false);
  const [getRuns, {data: runs = [], isLoading}] = useLazyGetRunListQuery();

  const handleOnClick = useCallback(() => {
    if (isCollapsed) {
      setIsCollapsed(false);
      return;
    }

    setIsCollapsed(true);
    TestAnalyticsService.onTestCardCollapse();
    if (runs.length > 0) {
      return;
    }
    getRuns({testId: test.id, take: 5});
  }, [getRuns, isCollapsed, runs.length, test.id]);

  return (
    <S.Container>
      <S.TestContainer onClick={() => handleOnClick()}>
        {isCollapsed ? <DownOutlined /> : <RightOutlined data-cy={`collapse-test-${test.id}`} />}
        <S.Box>
          <S.BoxTitle level={2}>{test.summary.runs}</S.BoxTitle>
        </S.Box>
        <S.TitleContainer>
          <S.Title level={3}>{test.name}</S.Title>
          <S.Text>
            {test.trigger.method} â€¢ {test.trigger.entryPoint}
          </S.Text>
        </S.TitleContainer>
        <div>
          <S.Text>Last run time:</S.Text>
          <Tooltip title={Date.format(test.summary.lastRun.time)}>
            <S.Text>{Date.getTimeAgo(test.summary.lastRun.time)}</S.Text>
          </Tooltip>
        </div>

        <div>
          <S.Text>Last run result:</S.Text>
          <S.Row>
            <Tooltip title="Passed assertions">
              <S.HeaderDetail>
                <S.HeaderDot $passed />
                {test.summary.lastRun.passes}
              </S.HeaderDetail>
            </Tooltip>
            <Tooltip title="Failed assertions">
              <S.HeaderDetail>
                <S.HeaderDot $passed={false} />
                {test.summary.lastRun.fails}
              </S.HeaderDetail>
            </Tooltip>
          </S.Row>
        </div>

        <S.Row>
          <S.RunButton
            type="primary"
            ghost
            data-cy={`test-run-button-${test.id}`}
            onClick={event => {
              event.stopPropagation();
              onRun(test.id);
            }}
          >
            Run
          </S.RunButton>
          <TestCardActions testId={test.id} onDelete={() => onDelete(test)} />
        </S.Row>
      </S.TestContainer>

      {isCollapsed && (
        <S.RunsContainer>
          {isLoading && (
            <S.LoadingContainer direction="vertical">
              <Skeleton.Input active block size="small" />
              <Skeleton.Input active block size="small" />
              <Skeleton.Input active block size="small" />
            </S.LoadingContainer>
          )}

          {Boolean(runs.length) && <ResultCardList testId={test.id} resultList={runs} />}

          {runs.length === 5 && (
            <S.FooterContainer>
              <S.Link data-cy="test-details-link" onClick={() => onViewAll(test.id)}>
                View all runs
              </S.Link>
            </S.FooterContainer>
          )}

          {!runs.length && !isLoading && (
            <S.EmptyStateContainer>
              <S.EmptyStateIcon />
              <S.Text disabled>No Runs</S.Text>
            </S.EmptyStateContainer>
          )}
        </S.RunsContainer>
      )}
    </S.Container>
  );
};

export default TestCard;
