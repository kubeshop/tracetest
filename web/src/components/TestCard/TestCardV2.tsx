import {DownOutlined, RightOutlined} from '@ant-design/icons';
import {Skeleton, Tooltip} from 'antd';
import {useCallback, useState} from 'react';

import {useLazyGetRunListQuery} from 'redux/apis/TraceTest.api';
import TestAnalyticsService from 'services/Analytics/TestAnalytics.service';
import {TTest} from 'types/Test.types';
import Date from 'utils/Date';
import * as S from './TestCardV2.styled';
import TestCardActions from './TestCardActions';
import ResultCardList from '../RunCardList/RunCardList';

interface IProps {
  onDelete(test: TTest): void;
  onRun(id: string): void;
  onViewAll(id: string): void;
  test: TTest;
}

const TestCardV2 = ({onDelete, onRun, onViewAll, test}: IProps) => {
  const [isCollapsed, setIsCollapsed] = useState(false);
  const [getRuns, {data: runs = [], isLoading}] = useLazyGetRunListQuery();

  const handleOnClick = useCallback(() => {
    if (isCollapsed) {
      setIsCollapsed(false);
      return;
    }

    setIsCollapsed(true);
    TestAnalyticsService.onTestCardCollapse();
    if (runs.length > 0) {
      return;
    }
    getRuns({testId: test.id, take: 5});
  }, [getRuns, isCollapsed, runs.length, test.id]);

  return (
    <S.Container>
      <S.TestContainer onClick={() => handleOnClick()}>
        <S.Row>
          {isCollapsed ? <DownOutlined /> : <RightOutlined />}
          <S.Box>
            <S.BoxTitle level={2}>0</S.BoxTitle>
          </S.Box>
          <div>
            <S.Title level={3}>{test.name}</S.Title>
            <S.Text>
              {test.trigger.method} â€¢ {test.trigger.entryPoint}
            </S.Text>
          </div>
        </S.Row>

        <S.Row $gap={36} $noWrap>
          <div>
            <S.Text>Last run time:</S.Text>
            <Tooltip title={Date.format('2022-09-26T17:17:22.064998211Z')}>
              <S.Text>{Date.getTimeAgo('2022-09-26T17:17:22.064998211Z')}</S.Text>
            </Tooltip>
          </div>

          <div>
            <S.Text>Last run result:</S.Text>
            <S.Row>
              <Tooltip title="Passed assertions">
                <S.HeaderDetail>
                  <S.HeaderDot $passed />0
                </S.HeaderDetail>
              </Tooltip>
              <Tooltip title="Failed assertions">
                <S.HeaderDetail>
                  <S.HeaderDot $passed={false} />0
                </S.HeaderDetail>
              </Tooltip>
            </S.Row>
          </div>

          <S.Row>
            <S.RunButton
              type="primary"
              ghost
              data-cy={`test-run-button-${test.id}`}
              onClick={event => {
                event.stopPropagation();
                onRun(test.id);
              }}
            >
              Run
            </S.RunButton>
            <TestCardActions testId={test.id} onDelete={() => onDelete(test)} />
          </S.Row>
        </S.Row>
      </S.TestContainer>

      {isCollapsed && (
        <S.RunsContainer>
          {isLoading && (
            <S.LoadingContainer direction="vertical">
              <Skeleton.Input active block size="small" />
              <Skeleton.Input active block size="small" />
              <Skeleton.Input active block size="small" />
            </S.LoadingContainer>
          )}

          {Boolean(runs.length) && <ResultCardList testId={test.id} resultList={runs} />}

          {runs.length === 5 && (
            <S.FooterContainer>
              <S.Link data-cy="test-details-link" onClick={() => onViewAll(test.id)}>
                View all runs
              </S.Link>
            </S.FooterContainer>
          )}

          {!runs.length && !isLoading && (
            <S.EmptyStateContainer>
              <S.EmptyStateIcon />
              <S.Text disabled>No Runs</S.Text>
            </S.EmptyStateContainer>
          )}
        </S.RunsContainer>
      )}
    </S.Container>
  );
};

export default TestCardV2;
