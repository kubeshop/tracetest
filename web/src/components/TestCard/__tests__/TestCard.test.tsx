import {fireEvent, render, waitFor} from '@testing-library/react';
import {ReduxWrapperProvider} from '../../../redux/ReduxWrapperProvider';
import {TestingModels} from '../../../utils/TestingModels';
import TestCard from '../TestCard';

test('TestCard', async () => {
  const onDelete = jest.fn();
  const onRunTest = jest.fn();
  const onClick = jest.fn();

  const {container, getByTestId} = render(
    <ReduxWrapperProvider>
      <TestCard onDelete={onDelete} onRunTest={onRunTest} test={TestingModels.test} onClick={onClick} />
    </ReduxWrapperProvider>
  );

  fireEvent(getByTestId(`test-actions-button-${TestingModels.test.testId}`), TestingModels.mouseEvent);
  await waitFor(() => getByTestId('test-card-delete'));
  expect(container).toMatchSnapshot();
});
