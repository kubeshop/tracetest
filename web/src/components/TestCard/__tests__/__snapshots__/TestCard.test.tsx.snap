// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestCard 1`] = `
<div>
  <div
    class="sc-jrQzAO guQEPO"
  >
    <div
      class="sc-kDTinF iNqRfS"
    >
      <span
        aria-label="right"
        class="anticon anticon-right"
        data-cy="collapse-test"
        role="img"
        tabindex="-1"
      >
        <svg
          aria-hidden="true"
          data-icon="right"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
          />
        </svg>
      </span>
      <div
        class="sc-crHmcD dqqtkd"
      >
        <span
          class="ant-typography sc-bqiRlB jCwMNy"
        >
          
        </span>
      </div>
      <div
        class="sc-crHmcD dqqtkd"
      >
        <span
          class="ant-typography sc-ksdxgE"
        >
          GET
        </span>
      </div>
      <div
        class="sc-crHmcD dqqtkd"
        data-cy="test-url-2309ei30"
      >
        <span
          class="ant-typography sc-ksdxgE"
        >
          
        </span>
      </div>
      <div
        class="sc-crHmcD dqqtkd"
      />
      <div
        class="sc-egiyK hbItou"
      >
        <button
          class="ant-btn ant-btn-primary ant-btn-background-ghost"
          data-cy="test-run-button-2309ei30"
          type="button"
        >
          <span>
            Run Test
          </span>
        </button>
      </div>
      <span
        class="ant-dropdown-trigger ant-dropdown-link ant-dropdown-open"
        data-cy="test-actions-button-2309ei30"
      >
        <span
          aria-label="more"
          class="anticon anticon-more sc-hBUSln"
          role="img"
          style="font-size: 24px; color: rgb(154, 163, 171); cursor: pointer;"
        >
          <svg
            aria-hidden="true"
            data-icon="more"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M456 231a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0zm0 280a56 56 0 10112 0 56 56 0 10-112 0z"
            />
          </svg>
        </span>
      </span>
    </div>
  </div>
</div>
`;
